### PR TITLE
[RDY] Pass line and column number to external editor

### LINKED ISSUE #2725: Pass line number to external editor
As @blueyed mentioned in IRC, it'd be nice if it was possible to pass the line number and maybe column to an external editor, if available.

### PR DESCRIPTION
**Pull request by 7lb**
Fixes #2725.

Added several possible placeholder to `editor.command` in order to be able to pass line and column numbers to the external editor in order to be able to place the cursor in the correct place in the buffer.

Added placegolders:

* `{file}` has the same function as `{}`
* `{line}` is the 1-based index of the target line in the buffer
* `{column}` is the 1-based index of the target column in the buffer
* `{line0}` is like `{line}` but 0-based
* `{column0}` is like `{column}` but 0-based

Also the placeholder check for shell commands accepts both `{}` or `{file}` placeholders

The reason `{line}` and `{column}` are 1-based by default is because most editors count lines and columns starting from 1. In fact I did not find any editor that starts counting lines or columns from 0.

Currently missing tests.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3100)

### CODE REVIEW

**Comment by 7lb** [comment]
I *think* the feature is mostly done, I just need to do some more testing regarding 0-width glyphs.

Speaking of testing, any help on how to add tests for this PR?

**Comment by 7lb** [comment]
Ok, after some more testing I must say there is some hidden complexity lying in this feature, but I'm not sure we can / should do something about that.

I only tested with gvim, which I guess is the main editor for qutebrowser's users, but I will test with other editors when I get a chance to see how they behave.

I found the following problems / gotchas:

* gvim usually counts byte offsets, not columns
* multi-byte characters visually occupy one column
* javascripts' `selectionStart` returns `the index of the first selected character`, which I *think* in this context really means unicode code-point.

The above means that setting the following `editor.command` will not correctly place the column marker in the presence of multi-byte characters:
```
'["gvim", "-c", "call cursor({line}, {column}", "-f", "{file}"]'
```

Fortunately, gvim has a motion that will move to a specified *virtual* column. Virtual column here means `column as visually rendered on the screen`. The motion is `N|` in normal mode. That is a number followed by a pipe.

I did not find a command to do the same (but my vim-foo is quite lacking beyond basic usage), but from the command line the following invocation works:
```
gvim -c "execute 'normal! {line}G {column}|'" -f {file}
```

Unfortunately the above has two problems:
1. I did not find a way to insert quotes in the arguments for `editor.command`, and this command requires them
2. tabs

Problem 1 is fixable, problem 2 less so. Basically the `N` in the `N|` motion does not represent visual *characters*, but visual *columns*. In the case of tabs the two will differ.

Moreover, javascript's `selectionStart` will (correctly) report 2 if the cursor is placed after two tabs in an otherwise empty line, but using the motion `2|` in gvim will **not** place the cursor "on" the second tab.

In fact it is impossible to tell what using the motion will do, because that depends on the `tabstop` setting of the gvim itself.

So, while a (I guess) viable hack could be to set `editor.command` to some kind of script that sets `tabstop=1`, then `execute 'normal! {line}G {column}|'` and then restores `tabstop`, that would mean that this funcionality would effectively be broken out-of-the-box for gvim users unless editing plain ascii text with no tabs.

That said, I don't really know what could be done qutebrowser-side.

Sorry, this turned out longer than expected.

**Comment by The-Compiler** [comment]
Thanks for the follow up! It might take me a bit until I get to looking at this, since I was really busy with the v1.0 release and the attention it has gotten over the past few days, so things started to pile up a bit.

File: qutebrowser/misc/editor.py [hunk 1]
```diff
@@ -96,11 +96,12 @@ def on_proc_closed(self, exitcode, exitstatus):
     def on_proc_error(self, _err):
         self._cleanup()
 
-    def edit(self, text):
+    def edit(self, text, caret_position):
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: This breaks other places calling `edit`, such as the `:edit-url` command (see [failing tests](https://travis-ci.org/qutebrowser/qutebrowser/jobs/287704068)). It should probably just default to 1.

File: qutebrowser/misc/editor.py [hunk 2]
```diff
@@ -121,21 +122,72 @@ def edit(self, text):
...
+        # For example in the text `aaa\nbb|bbb` (| represents the caret):
+        # caret_position = 6
+        # text[:caret_position] = `aaa\nbb`
+        # text[:caret_psotion].count('\n') = 1
+        # caret_position - text[:caret_position].rfind('\n') = 3
+        #
+        # Thus line, column = 2, 3, and the caret is indeed in the second
+        # line, third column
+        line = text[:caret_position].count('\n') + 1
+        column = caret_position - text[:caret_position].rfind('\n')
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: With the issues you've mentioned with getting the column, I feel like we should just drop and not support it. Probably nice enough to land on the right line already.

File: qutebrowser/misc/editor.py [hunk 3]
```diff
@@ -121,21 +122,72 @@ def edit(self, text):
...
+        #
+        # For example in the text `aaa\nbb|bbb` (| represents the caret):
+        # caret_position = 6
+        # text[:caret_position] = `aaa\nbb`
+        # text[:caret_psotion].count('\n') = 1
+        # caret_position - text[:caret_position].rfind('\n') = 3
+        #
+        # Thus line, column = 2, 3, and the caret is indeed in the second
+        # line, third column
+        line = text[:caret_position].count('\n') + 1
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: Please add some tests for that calculation to `tests/unit/misc/test_editor.py` (maybe easier if you move the caret calculation to its own function).

File: qutebrowser/misc/editor.py [hunk 4]
```diff
@@ -121,21 +122,72 @@ def edit(self, text):
...
         self._remove_file = False
         self._start_editor()
 
-    def _start_editor(self):
-        """Start the editor with the file opened as self._filename."""
+    def _start_editor(self, line=1, column=1):
+        """Start the editor with the file opened as self._filename.
+
+        Args:
+            caret_position: The position of the caret in the text.
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: Wrong docstring :wink:

File: qutebrowser/misc/editor.py [hunk 5]
```diff
@@ -121,21 +122,72 @@ def edit(self, text):
...
+        # text[:caret_psotion].count('\n') = 1
+        # caret_position - text[:caret_position].rfind('\n') = 3
+        #
+        # Thus line, column = 2, 3, and the caret is indeed in the second
+        # line, third column
+        line = text[:caret_position].count('\n') + 1
+        column = caret_position - text[:caret_position].rfind('\n')
+        self._start_editor(line=line, column=column)
 
     def edit_file(self, filename):
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: This now doesn't do any line number calculations - I feel like the calculation probably should go to `_start_editor` instead?

File: qutebrowser/misc/editor.py [hunk 6]
```diff
@@ -121,21 +122,72 @@ def edit(self, text):
...
+        Return:
+            The substituted placeholder or the original argument
+        """
+        sub = possible_placeholder\
+            .replace('{}', self._filename)\
+            .replace('{file}', self._filename)\
+            .replace('{line}', str(line))\
+            .replace('{line0}', str(line-1))\
+            .replace('{column}', str(column))\
+            .replace('{column0}', str(column-1))
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: Do something like this instead:

```python
replacements = {
    '{}': self._filename,
    '{file}': self._filename,
    ...
}

for old, new in replacements.items():
    possible_placeholder = possible_placeholder.replace(old, new)
```

File: qutebrowser/browser/webkit/webkitelem.py [hunk 7]
```diff
@@ -126,6 +126,14 @@ def set_value(self, value):
             value = javascript.string_escape(value)
             self._elem.evaluateJavaScript("this.value='{}'".format(value))
 
+    def caret_position(self):
+        self._check_vanished()
+        pos = self._elem.evaluateJavaScript('this.selectionStart')
+        assert isinstance(pos, (int, float, type(None)))
```
[the remark this answers is not in the document]
Comment (7lb) [reply qutebrowser/browser/webkit/webkitelem.py]: We agree it should be an int, but for some reason it's always a float 😄

File: qutebrowser/browser/webkit/webkitelem.py [hunk 8]
```diff
@@ -126,6 +126,14 @@ def set_value(self, value):
...
+        if pos is None:
+            return 0
+        return int(pos)
```
[the remark this answers is not in the document]
Comment (7lb) [reply qutebrowser/browser/webkit/webkitelem.py]: As stated, it surprisingly is a float

File: qutebrowser/misc/editor.py (hunk 2 above)
Comment (7lb) [reply qutebrowser/misc/editor.py]: I'm not sure about this, the calculation is correct, provided the column number represents unicode code-points. While the feature is broken for the vi family of editors I feel it will probably work correctly with editors such as atom or sublime text, but I have yet to test this.

In any case, I can remove it if you feel that's the best course of action.

File: qutebrowser/misc/editor.py (hunk 5 above)
Comment (7lb) [reply qutebrowser/misc/editor.py]: Yes, I initially placed the code in `_start_editor` (as you can see by the wrong docstring) but than realized I needed the text to calculate the `line` and `column`.

For some reason it feels cleaner to pass `line` and `column` to `_start_editor` rather than `caret_position` and `text` but I don't really know why.

In any case, if you feel it's better to calculate (or rather call a function that calculates) those values inside `_start_editor` I will proceed to do that 😃

File: qutebrowser/misc/editor.py (hunk 2 above)
Comment (The-Compiler) [reply qutebrowser/misc/editor.py]: I think it depends whether other editors actually have a way to set the cursor column when invoking. The problem with having it but it not working correctly with vim is that I get to deal with the bug reports then :wink:

Do you know whether the cursor can be set like this in other editors?

File: qutebrowser/misc/editor.py [hunk 9]
```diff
@@ -121,21 +122,93 @@ def edit(self, text):
...
+        """Substitute a single placeholder.
+
+        If the `possible_placeholder` input to this function is valid it will
+        be substituted with the appropriate value, otherwise it will be left
+        unchanged.
+
+        Args:
+            possible_placeholder: an argument of editor.command.
+            line: the previously-calculated line number for the text caret
+            column: the previously-calculated column number for the text caret
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: nitpick: Please add a period (here and in the line above).

File: qutebrowser/misc/editor.py [hunk 10]
```diff
@@ -121,21 +122,93 @@ def edit(self, text):
...
+
+        Return:
+            The substituted placeholder or the original argument
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: nitpick: Please add a period.

**Comment by Ape** [comment]
@7lb I think this motion works in vim: `normal {line}gg0{column}l`

File: qutebrowser/misc/editor.py [hunk 11]
```diff
@@ -121,21 +122,93 @@ def edit(self, text):
...
+        # the last newline before the caret.
+        #
+        # For example in the text `aaa\nbb|bbb` (| represents the caret):
+        # caret_position = 6
+        # text[:caret_position] = `aaa\nbb`
+        # text[:caret_psotion].count('\n') = 1
+        # caret_position - text[:caret_position].rfind('\n') = 3
+        #
+        # Thus line, column = 2, 3, and the caret is indeed in the second
+        # line, third column
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: You could probably move that comment to the docstring now (before `Args:`).

File: qutebrowser/misc/editor.py [hunk 12]
```diff
@@ -121,21 +122,93 @@ def edit(self, text):
...
+        # To find the line we just count how many newlines there are before the
+        # caret and add 1.
+        #
+        # To find the column we calculate the difference between the caret and
+        # the last newline before the caret.
+        #
+        # For example in the text `aaa\nbb|bbb` (| represents the caret):
+        # caret_position = 6
+        # text[:caret_position] = `aaa\nbb`
+        # text[:caret_psotion].count('\n') = 1
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: psotion -> position

File: qutebrowser/misc/editor.py [hunk 13]
```diff
@@ -121,21 +122,93 @@ def edit(self, text):
...
+            '{}': self._filename,
+            '{file}': self._filename,
+            '{line}': str(line),
+            '{line0}': str(line-1),
+            '{column}': str(column),
+            '{column0}': str(column-1)
+        }
+
+        # The for loop would create a copy anyway, this should be more legible
+        ph = possible_placeholder
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: Maybe just rename the `possible_placeholder` argument to `arg` (as in editor argument) or so? After all, it can also be `foo{placeholder}bar` or so.

(Also, note that neither is a copy - just [different references to the same object](http://python.net/~goodger/projects/pycon/2007/idiomatic/handout.html#other-languages-have-variables) :wink:)

File: tests/unit/config/test_configtypes.py [hunk 14]
```diff
@@ -1818,6 +1818,9 @@ def klass(self):
         ({'placeholder': '{}'}, '[foo, "{}", bar]', ['foo', '{}', 'bar']),
         ({'placeholder': '{}'}, '["foo{}bar"]', ['foo{}bar']),
         ({'placeholder': '{}'}, '[foo, "bar {}"]', ['foo', 'bar {}']),
+        ({'placeholder': '{file}'}, '[f, "{file}", b]', ['f', '{file}', 'b']),
+        ({'placeholder': '{file}'}, '["f{file}b"]', ['f{file}b']),
+        ({'placeholder': '{file}'}, '[f, "b {file}"]', ['f', 'b {file}']),
```
Comment (The-Compiler) [inline tests/unit/config/test_configtypes.py]: Hmm, now my own tests got me confused. In `configtypes.py`, `placeholders` is just a boolean, so this should probably actually be `True` instead of `'{}'` or `'{file}'`.

File: tests/unit/misc/test_editor.py [hunk 15]
```diff
@@ -157,6 +157,24 @@ def test_double_edit(self, editor):
             editor.edit("")
 
 
+class TestLineColumnCalculation:
```
Comment (The-Compiler) [inline tests/unit/misc/test_editor.py]: No need for a class here if you just have a single test method - just make that a function instead, similar to `test_modify`.

**Comment by 7lb** [comment]
Thanks for the feedbacks, I will address them this evening after work 👍

**Comment by 7lb** [comment]
@Ape wow. That was so simple I almost feel stupid! Definitely works! 😄 
Also doesn't need those pesky quotes-inside-quotes since `:normal` is apparently a vim command.

The full command line would thus be `gvim -f {file} -c "normal {line}G{column0}l"` and that'll work with multi-byte, zero-width and tab characters :)

Of course, one'll need to change that if `G` or `l` are remapped.

I'll test non-vim editors ASAP to see how they behave

File: qutebrowser/misc/editor.py (hunk 2 above)
Comment (7lb) [reply qutebrowser/misc/editor.py]: Thanks to Ape's suggestion we have a working command for vim and family.

I also tested atom and sublime text 3, both have no problems with multi-byte or zero-width characters, and both need the same command `'["atom", "{file}:{line}:{column}"]'`

**Comment by mschilli87** [comment]
@7lb: I don't know from the top of my head but I guess it would even be possible to adjust this to a `--vanilla`/`--no-rc` `[g]vim` call (making sure `G` & `l` are not re-mapped), executing the motion and explicitely loading user config after that.

**Comment by The-Compiler** [comment]
Can you please take a look at [the failing tests](https://travis-ci.org/qutebrowser/qutebrowser/jobs/289654633#L584), [pylint output](https://travis-ci.org/qutebrowser/qutebrowser/jobs/289654643#L482) and [flake8 output](https://travis-ci.org/qutebrowser/qutebrowser/jobs/289654644)?

File: doc/help/settings.asciidoc [hunk 16]
```diff
@@ -1913,7 +1913,14 @@ Default: +pass:[-1]+
...
 The editor (and arguments) to use for the `open-editor` command.
-`{}` gets replaced by the filename of the file to be edited.
+Several placeholders are supported. Placeholders are substituted by the
+respective value when executing the command.
+
+`{file}` gets replaced by the filename of the file to be edited.
+`{line}` gets replaced by the line in which the caret is found in the text.
+`{column}` gets replaced by the column in which the caret is found in the text.
+`{line0}` same as `{line}`, but starting from index 0
+`{column0}` same as `{column}`, but starting from index 0
```
Comment (The-Compiler) [inline doc/help/settings.asciidoc]: Please add trailing periods for consistency with the lines above.

**Comment by 7lb** [comment]
Okay I fixed the broken tests (the default config has changed) and added docstrings to the added `caret_position` methods, along with the missing full-stops.

The tests pass, pylint is slighlty happied but still fails because there is bad whitespace and other missing docstrings, but imo those changes fall outside of this PR.

I can try and fix those none the less, just let me know :)

Also, I'm marking this as ready since I believe the functionality to be complete.

**Comment by The-Compiler** [comment]
pylint seems to pass on CI (against the latest master) - however, it looks like [the tests fail with QtWebEngine still](https://travis-ci.org/qutebrowser/qutebrowser/jobs/291205056#L521) and [flake8 also has some complaints](https://travis-ci.org/qutebrowser/qutebrowser/jobs/291205065#L481).

**Comment by 7lb** [comment]
I'll look into it as soon as I can. flake8 seemed to pass on my machine.

Sorry for the delays.



24 Oct 2017, 08:29 by notifications@github.com:

**Comment by 7lb** [comment]
@The-Compiler again sorry for the delays, I planned on looking at this yesterday but life happened.

In any case, I fixed the flake8 failures, but cannot reproduce the test failures in `test_editor_bdd.py`, in fact looking at the travis log it looks like some sort of javascript timeout.

It doesn't happen on my machine, could it be that travis does something different that triggers the timeout?

**Comment by The-Compiler** [comment]
No worries! It looks like it's only happening with QtWebEngine, so you'll need to run tox with `-- --qute-bdd-webengine` to reproduce it.

Looking at the log, this is the relevant bit:

```
19:20:33.119 DEBUG    js         shared:javascript_log_message:136 [:134] Uncaught InvalidStateError: Failed to read the 'selectionStart' property from 'HTMLInputElement': The input element's type ('button') does not support selection.
19:20:33.120 DEBUG    webview    webenginetab:_js_cb_single:474 Got element from JS: None
19:20:33.121 ERROR    message    message:error:63 No element found with id qute-button!
```

This is when clicking a button, after editing the text:

```gherkin
Feature: Opening external editors
  Scenario: Spawning an editor successfully (features/editor.feature:63)
    When I set up a fake editor returning "foobar" (0.73s)
    And I open data/editor.html (0.16s)
    And I run :click-element id qute-textarea (0.04s)
    And I wait for "Entering mode KeyMode.insert (reason: clicking input)" in the log (0.01s)
    And I run :open-editor (0.05s)
    And I wait for "Read back: foobar" in the log (0.05s)
    And I run :click-element id qute-button (0.03s)
    Then the javascript message "text: foobar" should be logged (14.90s)
```

**Comment by 7lb** [comment]
Oh, my bad! I thought webengine was the default backend during tests too.

Yeah I can reproduce it now, I'm looking into it.

**Comment by 7lb** [comment]
Ok, I added a `try {...} catch {...}` to fix behaviour. Tests pass now.

I guess webkit returns undefined instead of throwing an exception, and that's why it was broken only in webengine.

File: qutebrowser/javascript/webelem.js [hunk 17]
```diff
@@ -48,11 +48,22 @@ window._qutebrowser.webelem = (function() {
...
         elements[id] = elem;
 
+        // InvalidStateError will be thrown if elem doesn't have selectionStart
+        var caret_position = 0;
+        try {
+            caret_position = elem.selectionStart;
+        } catch (e) {
+            if (e instanceof DOMException && e.name === "InvalidStateError") {
+                // nothing to do, caret_position is already 0
+            }
```
Comment (The-Compiler) [inline qutebrowser/javascript/webelem.js]: otherwise you should probably throw it?

**Comment by The-Compiler** [comment]
I think the reason it happened is that QtWebKit only runs `selectionStart` on elements when it was actually requested for that element, while QtWebEngine does so for *all* elements to be serialized, even if it's not an input.

File: qutebrowser/javascript/webelem.js (hunk 17 above)
Comment (7lb) [reply qutebrowser/javascript/webelem.js]: ... Of course 🙄

**Comment by 7lb** [comment]
I see, yeah that would do it.

I should have fixed all my silly mistakes 😄 
I hope so anyway

**Comment by jgkamat** [comment]
I think tests are passing, but I wanted to let you know that the js linter is throwing a couple warnings:

```
/home/travis/build/qutebrowser/qutebrowser/qutebrowser/javascript/webelem.js
  55:18  error  Identifier name 'e' is too short (< 2)  id-length
  60:24  error  Missing semicolon                       semi
```

File: qutebrowser/browser/webkit/webkitelem.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (NoctuaNivalis) [reply qutebrowser/browser/webkit/webkitelem.py]: If we convert it to an `int` anyway, why'd you check the type first? Using
```python
pos = self._elem.evaluateJavaScript('this.selectionStart')
return int(pos or 0)
```
should be equivalent for most types, except it'll throw a ValueError instead of an AssertionError. Furthermore, it'll also accept other numerical types.

[the remark this answers is not in the document]
Comment (The-Compiler) [reply qutebrowser/browser/webkit/webkitelem.py]: I agree we can get rid of the `isinstance` check here, but I'd rather not do `int(pos or 0)` - I've seen too many funny things happen with JavaScript bridged to C++. If something goes wrong here, and Qt returns `{}` or something, I'd rather know about it than just assume 0.

**Comment by 7lb** [comment]
Ok, I fixed the eslint warnings. I am not sure on the recommended style for multine if conditions though.

In any case, sorry for all the formatting and linting errors, but there are many on my local branch and can be difficult to distinguish the signal from the noise for me ^^'

File: qutebrowser/browser/webkit/webkitelem.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (7lb) [reply qutebrowser/browser/webkit/webkitelem.py]: I'm sorry, I am having trouble understanding what is requested here. I gather the `isinstance` check can be removed, but at that point what is the difference between

```python
if pos is None:
    return 0
return int(pos)
```

and 

```python
return int(pos or 0)
```

?

Maybe I'm misunderstanding and I should in fact leave `isinstance` alone?

[the remark this answers is not in the document]
Comment (The-Compiler) [reply qutebrowser/browser/webkit/webkitelem.py]: The difference between the two is how e.g. `{}` is handled. I've seen QtWebEngine just return `{}` when accidentally passing some kind of JS object which isn't just made up of "primitive types".

If we do something like that, I'd much rather get an exception here than just silently assuming 0 (the reason I originally added those type checks is that I spent quite some time debugging a funny issue like that).

With

```python
if pos is None:
    return 0
return int(pos)
```

passing `{}` fails like expected, while with `return int(pos or 0)`, `0` is used.

File: qutebrowser/javascript/webelem.js [hunk 18]
```diff
@@ -48,11 +48,28 @@ window._qutebrowser.webelem = (function() {
...
 
+        // InvalidStateError will be thrown if elem doesn't have selectionStart
+        var caret_position = 0;
+        try {
+            caret_position = elem.selectionStart;
+        } catch (err) {
+            if (
+                err instanceof DOMException &&
+                err.name === "InvalidStateError"
+            ) {
```
Comment (The-Compiler) [inline qutebrowser/javascript/webelem.js]: Coming from Python, I'd format this like:

```javascript
            if (err instanceof DOMException &&
                    err.name === "InvalidStateError") {
                // nothing to do, caret_position is already 0
```

File: qutebrowser/misc/editor.py [hunk 19]
```diff
@@ -121,21 +122,92 @@ def edit(self, text):
...
+        Args:
+            text: the text for which the numbers must be calculated
+            caret_position: the position of the caret in the text
+
+        Return:
+            A (line, column) tuple of (int, int)
+        """
+        line = text[:caret_position].count('\n') + 1
+        column = caret_position - text[:caret_position].rfind('\n')
+        return (line, column)
```
Comment (The-Compiler) [inline qutebrowser/misc/editor.py]: nitpick: No need for the parens here.

**Comment by The-Compiler** [comment]
> In any case, sorry for all the formatting and linting errors, but there are many on my local branch and can be difficult to distinguish the signal from the noise for me ^^'

No worries! I commented on some more things, and then this should be ready to merge.

You're probably seeing a lot of `multiline-comment-style` errors? This is because a new eslint version came out with a new checker - qutebrowser disabled that in 5dacf1431f4d3f15afc232938b5a7feea71e5b33 but that's not in your branch. You can just apply that temporarily (but not commit it) to get a better overview of errors hopefully.

File: qutebrowser/browser/webkit/webkitelem.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (7lb) [reply qutebrowser/browser/webkit/webkitelem.py]: I see, it makes sense. I had tought that empty stuff would generally map to `None` but I held the wrong assumption.

**Comment by The-Compiler** [comment]
Thanks!

**Comment by The-Compiler** [comment]
I haven't noticed this earlier, but it looks like you edited `settings.asciidoc` directly - however, that is generated from `configdata.yml`. I fixed it in https://github.com/qutebrowser/qutebrowser/commit/5469a238de3a21635a6aea068c8e61ad8da11def

**Comment by 7lb** [comment]
Oops! Awfully sorry, I didn't see the warning / disclaimer at the top of the file ☹️

**Comment by Ape** [comment]
That's why it's not advised to have any generated files in version control, but instead automatically generate the files from the source files.

**Comment by The-Compiler** [comment]
@Ape That's the reason why the generated HTML is not in the repo :wink: Not having the asciidoc files there has more drawbacks than advantages, which is why they are there.
